Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs

```diff
@@ -72,4 +332,67 @@ postUiWmbTripEnd ::
     API.Types.UI.WMB.TripEndReq ->
     Environment.Flow API.Types.UI.WMB.TripEndResp
   )
-postUiWmbTripEnd = do error "Logic yet to be decided"
+postUiWmbTripEnd (person, _, merch_city_id) tripTransactionId obj = do
+  driverId <- fromMaybeM (InternalError "Driver ID not found") person
+  fleetDriverAssociation <- FDA.findByDriver driverId >>= fromMaybeM (InternalError "Fleet Association not found")
+  transaction <- QTT.findByTransactionId (Kernel.Types.Id.Id tripTransactionId) >>= fromMaybeM (InternalError "no trip transaction found")
+  routeInfo <- QRM.findByRouteCode transaction.routeCode >>= fromMaybeM (InternalError "Route not found")
+  transporterConfig <- QTC.findByMerchantOpCityId merch_city_id >>= fromMaybeM (InternalError "Config not found")
+  fleetConfigInfo <- QFC.findByPrimaryKey (Kernel.Types.Id.Id fleetDriverAssociation.fleetOwnerId) >>= fromMaybeM (InternalError "Fleet Config Info not found")
+  now <- getCurrentTime
+
+  let distanceLeft = KU.distanceBetweenInMeters obj.location routeInfo.endPoint
+      updateAndRespond response = do
+        QDI.updateOnRide False driverId
+        QTT.updateStatus COMPLETED (Just routeInfo.endPoint) (Kernel.Types.Id.Id tripTransactionId)
+        TN.notifyWmbOnRide driverId merch_city_id COMPLETED "Ride Ended" "Your ride has ended" Nothing
+        pure response
+
+      createDriverRequest = do
+        requestId <- generateGUID
+        let info =
+              WmbEndTripData
+                { lat = obj.location.lat,
+                  lon = obj.location.lon,
+                  reason = Nothing,
+                  tripTransactionId = Kernel.Types.Id.Id tripTransactionId
+                }
+        let driver_req =
+              ApprovalRequest
+                { id = requestId,
+                  requestType = WmbEndTrip info,
+                  title = "Your ride has ended",
+                  body = "You reached your end stop of this route",
+                  status = PENDING,
+                  reason = Nothing,
+                  createdAt = now,
+                  updatedAt = now,
+                  merchantId = Nothing,
+                  merchantOperatingCityId = Just merch_city_id
+                }
+        QDR.create driver_req
+        pure $ TripEndResp {requestId = Just requestId.getId, result = WAITING_FOR_ADMIN_APPROVAL}
+
+  if fleetConfigInfo.allowEndingMidRoute || distanceLeft <= transporterConfig.endRideDistanceThreshold
+    then updateAndRespond $ TripEndResp {requestId = Nothing, result = SUCCESS}
+    else
+      if fleetConfigInfo.rideEndApproval
```

Comment:
`rideEndApproval` not clear if this means its required or notRequired, can be mistaken later.